### ISSUE TITLE
Auto-completion doesnt work for unimported packages

### ISSUE THREAD

**Issue by ramya-rao-a** [Microsoft/vscode-go#1944]
Using https://github.com/stamblerre/gocode

In the below case we dont get any completions.
```
import (
        "rsc.io/quote"
        "fmt"
)

func main() {
        fmt.Println("hello")
        quote.
}
```

Update the `quote.` above to `quote.Hello()`, save the file. The next time auto-completions work

```
import (
        "rsc.io/quote"
        "fmt"
)

func main() {
        fmt.Println("hello")
        quote.Hello();
        quote.
}
```

@stamblerre Logging this issue here for tracking purposes as your fork doesnt support logging issues

**Comment by stamblerre** [comment]
@ramya-rao-a thanks for reporting this. I just enabled issues and copied it to https://github.com/stamblerre/gocode/issues/3.

**Comment by dreamheld** [comment]
You, guys, are awesome keeping on top of these things the way you do!!!

**Comment by ramya-rao-a** [comment]
@stamblerre Thanks! I'll keep this issue open because invariably, folks will log an issue in this repo.

**Comment by JoseFMP** [comment]
To be fair, it also happens with imported packages.

**Comment by joaoferrao** [comment]
@stamblerre I tried installing according to your suggestion, restarting VSCode but auto completion still doesn't happen if the project is outside the $GOPATH.

**Comment by stamblerre** [comment]
@joaoferrao Did VSCode prompt you to install the gocode-gomod tool, and did you install it? There is a separate version of gocode that works specifically for modules.

**Comment by joaoferrao** [comment]
@stamblerre first time around yes, but I didn't know I should also install beta version of the extension. Then I reinstalled VSCode, installed beta version but wasn't asked to install the fork.

**Comment by stamblerre** [comment]
@joaoferrao: Yep, just execute the following commands:

`go get -u -d github.com/stamblerre/gocode`
`go build -o $GOPATH/bin/gocode-gomod github.com/stamblerre/gocode`

You should end up with a tool `gocode-gomod` in your `$GOPATH/bin`.

**Comment by joaoferrao** [comment]
@stamblerre thank you for getting back so fast, but I still am in no luck. I still only see the "PANIC".

![image](https://user-images.githubusercontent.com/28150182/46965287-b9802180-d0aa-11e8-993d-639d48a41170.png)

I also confirm I have the beta version of the Go extension:
![image](https://user-images.githubusercontent.com/28150182/46965321-dc123a80-d0aa-11e8-837c-24fe2443011d.png)

Is there anything else I can check? Any setting I should enforce in the settings.json for the go extension?

**Comment by stamblerre** [comment]
@joaoferrao I'd recommend running `gocode close` and `gocode-gomod close` to restart gocode, just to be safe. Then, run `gocode-gomod -s -debug` and try to complete in VSCode. The debug output should be printed there, and if you paste it in here, I can try to investigate it.

**Comment by joaoferrao** [comment]
@stamblerre outstanding support, thanks!! Maybe would be worthwhile

![image](https://user-images.githubusercontent.com/28150182/46965841-35c73480-d0ac-11e8-96a6-dafe32b986f0.png)

So for others who end up here with the same issue:

```bash
cd /tmp
curl -O https://github.com/Microsoft/vscode-go/releases/download/latest/Go-0.6.92-beta.1.vsix
code --install-extension Go-latest.vsix
go get -u -d github.com/stamblerre/gocode
go build -o $GOPATH/bin/gocode-gomod github.com/stamblerre/gocode
gocode close
gocode-gomod close
```

Kill VSCode and relaunch.

Thanks again.

**Comment by stamblerre** [comment]
@joaoferrao awesome, glad it works!

**Comment by zeeraw** [comment]
I want to confirm that the package is found

![Package is found](https://puu.sh/BNzee/430a7cb61a.png)

But once I add a dot to autocomplete the package, nothing is returned

![](https://puu.sh/BNzet/a3111ad7b5.png)

I go back to step one to manually import the package, there is no autocompletion on the package

![](https://puu.sh/BNzeQ/4dea8ff42a.png)

Not until I save the file with the import added, will I get an autocomplete list for my package

![](https://puu.sh/BNzfx/bc40bf2acd.png)

Hope this helps! Keep up the good work.

**Comment by stamblerre** [comment]
@zeeraw yep, this issue is tracked in https://github.com/stamblerre/gocode/issues/3. It's actually the result of a problem in the go/packages API that still needs to be solved. I will update on the issue once the go/packages API has fixed this problem.

**Comment by joaoferrao** [comment]
@stamblerre so the go extension decided to update from beta to the latest release and clearly is not behaving correctly (the beta seemed more stable).

I tried repeating the process for installing the beta versions .vsix files but:
```bash
 $  code --install-extension Go-0.6.94-beta.5.vsix
Corrupt ZIP: end of central directory record signature not found
```

**Comment by stamblerre** [comment]
@joaoferrao, I think this is probably a question for @ramya-rao-a?

**Comment by joaoferrao** [comment]
OK... just found out it's a problem related of downloading it with curl...

**Comment by nezorflame** [comment]
Another observation: behavior is different being inside $GOPATH with `"go.toolsEnvVars": {"GO111MODULE": "on"}` and outside $GOPATH without env.
In the second case, autocomplete indeed works after the package was used once:
![image](https://user-images.githubusercontent.com/3906766/48307684-19130500-e563-11e8-9ee7-0a18456bcedb.png)
The result of the function is also recognized well:
![image](https://user-images.githubusercontent.com/3906766/48307689-30ea8900-e563-11e8-9053-b13a5b604348.png)

**Comment by stamblerre** [comment]
@nezorflame: in the case where "id" had an invalid type, there appears to be a red underline beneath the import, so I wonder if that's related to the issue. 

I'm also not sure why gocode would behave differently depending on the location of the code, but I would guess that it's not switching between the module-unaware and module-aware versions of gocode correctly in these cases. VSCode-Go checks that value of `go env GOMOD` to decide which tool to use, so if you have `GO111MODULE=on` set, then it should work correctly. @ramya-rao-a, what do you think?

**Comment by nezorflame** [comment]
@stamblerre after some research, it seems that if the file's not saved yet, these `invalid type` and `could not import...` errors happen.

So, basically, you can't have autocomplete, auto-import etc. until your code until the file is saved.

**Comment by gencer** [comment]
@ramya-rao-a, save or not. Anything outside of current file does not auto-complete. Only internal (core) is getting autocompleted. (Like ioutil, strings...)

**Comment by stamblerre** [comment]
@ramya-rao-a: I am pretty sure that I can reproduce the issue, though my VSCode-Go settings are very strange at this point so it's possible I'm missing something. When I set `GO111MODULE=on` through my terminal, it started working, so I expect it must be something to do with propagating the `go.ToolsEnvVars`.

@nezorflame: your issue is known and tracked by https://github.com/stamblerre/gocode/issues/3.

 @gencer: is this related to the bug you already filed here https://github.com/mdempsky/gocode/issues/84? Are you using modules? [mdempsky/gocode](https://github.com/mdempsky/gocode) does not support modules, so if you are experiencing issues with completion using Go modules, then it would be best to file a separate issue under [stamblerre/gocode](https://github.com/stamblerre/gocode).

**Comment by ramya-rao-a** [comment]
@gencer @nezorflame Can you try running `gocode-gomod close`, ensure there are no running instances of it and try again?

Once gocode is running in server mode, the same is re-used everytime we ask for suggestions. It may happen the current running instance isnt aware of the GO111MODULE env var and has to be restarted?

**Comment by gencer** [comment]
@ramya-rao-a, My issue is about generic autocomplete but not modules (so no *-gomod running). I will continue this issue on other repository. For now, I will follow there.

**Comment by nezorflame** [comment]
@ramya-rao-a just checked.

- inside `$GOPATH` with `"go.toolsEnvVars": {"GO111MODULE": "on"}`:

```bash
~ ps ax | grep gocode | grep -v grep
1454   ??  S      0:00.04 /Users/nezorflame/Go/bin/gocode -s -sock unix -addr 127.0.0.1:37373
```

- outside `$GOPATH` with and without `"go.toolsEnvVars": {"GO111MODULE": "on"}`:

```bash
~ ps ax | grep gocode | grep -v grep
5108   ??  S      0:00.08 /Users/nezorflame/Go/bin/gocode-gomod -s -sock unix -addr 127.0.0.1:37373
```

**Comment by shaodahong** [comment]
when i run
```bash
~/go/bin
0² ./gocode close
2018/11/26 21:47:19 dial unix /var/folders/j3/1mmzgr1j51s_kgdg28ksn7b40000gn/T/gocode-daemon.shaodahong: connect: no such file or directory
```

**Comment by stamblerre** [comment]
@shaodahong: that just means that your gocode server wasn't running.

**Comment by shaodahong** [comment]
@stamblerre I know when I coding in vscode , gocode start run, but autocomplete still bad, I have some question

1.
if run `gocode-gomod -s -debug` a part autocomplete is good, but when I `ctrl+c` exit, autocomplete bad

2.
if run `gocode -s -debug`, no log display

go env

```bash
0² go env
GOARCH="amd64"
GOBIN=""
GOCACHE="/Users/shaodahong/Library/Caches/go-build"
GOEXE=""
GOFLAGS=""
GOHOSTARCH="amd64"
GOHOSTOS="darwin"
GOOS="darwin"
GOPATH="/Users/shaodahong/go"
GOPROXY=""
GORACE=""
GOROOT="/usr/local/go"
GOTMPDIR=""
GOTOOLDIR="/usr/local/go/pkg/tool/darwin_amd64"
GCCGO="gccgo"
CC="clang"
CXX="clang++"
CGO_ENABLED="1"
GOMOD="/Users/shaodahong/Workspace/demo/go.mod"
CGO_CFLAGS="-g -O2"
CGO_CPPFLAGS=""
CGO_CXXFLAGS="-g -O2"
CGO_FFLAGS="-g -O2"
CGO_LDFLAGS="-g -O2"
PKG_CONFIG="pkg-config"
GOGCCFLAGS="-fPIC -m64 -pthread -fno-caret-diagnostics -Qunused-arguments -fmessage-length=0 -fdebug-prefix-map=/var/folders/j3/1mmzgr1j51s_kgdg28ksn7b40000gn/T/go-build051668293=/tmp/go-build -gno-record-gcc-switches -fno-common"
```

**Comment by stamblerre** [comment]
@shaodahong: just to clarify, are you having the same problem as @nezorflame? When you open a file in a package that uses modules, which version of gocode gets started? Does `gocode-gomod close` work?

**Comment by shaodahong** [comment]
new log
```bash
0² gocode-gomod -s -debug
2018/11/27 12:29:43 Got autocompletion request for '/Users/shaodahong/Workspace/demo/main.go'
2018/11/27 12:29:43 Cursor at: 30
2018/11/27 12:29:43 -------------------------------------------------------
package main

func main() {
	f#
}
2018/11/27 12:29:43 -------------------------------------------------------
2018/11/27 12:29:44 Elapsed duration: 195.027647ms
2018/11/27 12:29:44 Offset: 0
2018/11/27 12:29:44 Number of candidates found: 3
2018/11/27 12:29:44 Candidates are:
2018/11/27 12:29:44   const false untyped bool
2018/11/27 12:29:44   type float32 float32
2018/11/27 12:29:44   type float64 float64
2018/11/27 12:29:44 =======================================================
2018/11/27 12:29:44 Got autocompletion request for '/Users/shaodahong/Workspace/demo/main.go'
2018/11/27 12:29:44 Cursor at: 33
2018/11/27 12:29:44 -------------------------------------------------------
package main

func main() {
	fmt.#
}
2018/11/27 12:29:44 -------------------------------------------------------
2018/11/27 12:29:44 Elapsed duration: 114.33626ms
2018/11/27 12:29:44 Offset: 0
2018/11/27 12:29:44 Number of candidates found: 0
2018/11/27 12:29:44 Candidates are:
2018/11/27 12:29:44 =======================================================
2018/11/27 12:29:44 Got autocompletion request for '/Users/shaodahong/Workspace/demo/main.go'
2018/11/27 12:29:44 Cursor at: 46
2018/11/27 12:29:44 -------------------------------------------------------
package main
import "fmt"

func main() {
	fmt.#
}
2018/11/27 12:29:44 -------------------------------------------------------
2018/11/27 12:29:44 Elapsed duration: 113.933882ms
2018/11/27 12:29:44 Offset: 0
2018/11/27 12:29:44 Number of candidates found: 0
2018/11/27 12:29:44 Candidates are:
2018/11/27 12:29:44 =======================================================

```

**Comment by stamblerre** [comment]
@shaodahong: what changed between this and your other comment (https://github.com/Microsoft/vscode-go/issues/2148#issuecomment-441901878)?

**Comment by nezorflame** [comment]
Gave HEAD version another go (updated with vscode-go feature `Install/Update Tools`).

Piece of code:
```go
package main

type A struct {
	S string
}

func main() {
	a := A{}
}
```

Saved outside `$GOPATH` at `$HOME/Go/test-project/main.go`, so modules should automatically work.

Preparation:
- `gocode close`
- `gocode-gomod close`
- `go mod init test-project`
- `gocode-gomod -s -debug`

Tests:

0. **Action:** Entering `fm`.
**Expected:** autocompletion for `fmt` package, on acceptance `fmt` is imported.
**Result:** works:
```go
package main

import (
	"fmt"
)

type A struct {
	S string
}

func main() {
	a := A{}
	fmt
}
```
1. **Action:** Entering `.` after `fmt`
**Expected:** functions of `fmt` package
**Result:** nothing:
![image](https://user-images.githubusercontent.com/3906766/49097643-34c61d00-f27e-11e8-9cf1-b84af0daaf39.png)
**Output**: <https://pastebin.com/GVCH02Em>
2. **Action:** Saving the file, deleting and re-entering `.` after `fmt`
**Expected:** functions of `fmt` package should appear now
**Result:** works:
![image](https://user-images.githubusercontent.com/3906766/49097977-eebd8900-f27e-11e8-9e11-1d7746f6c369.png)
**Output**: <https://pastebin.com/PjpDU2nA>
3. **Action:** Entering `Println(a.)` after `fmt.`, cursor at `.`
**Expected:** struct field `S` should be shown in autocomplete
**Result:** works:
![image](https://user-images.githubusercontent.com/3906766/49098163-51168980-f27f-11e8-98ba-2658c8ccd845.png)
**Output**: <https://pastebin.com/PyRSzKPa>

**Comment by shaodahong** [comment]
@stamblerre new log I delete `import "fmt"`

**Comment by shaodahong** [comment]
`GO111MODULE ` i think no need set, go version 1.11 can auto judge it, when outside GOPATH and have go.mod, go module become effective

**Comment by shaodahong** [comment]
Maybe you can refer to the `intellij IDEA` go plugin, I don't know is the same as VScode.
![image](https://user-images.githubusercontent.com/14831261/49123639-506c0a80-f2f4-11e8-8808-7fb1d9347c91.png)

**Comment by nezorflame** [comment]
@shaodahong you have to force `GO111MODULE=on` if you're inside `$GOPATH` and want to use modules (which is my current scenario and the preferred one, 'cause I like the idea of `$GOPATH`).

Regarding the IDEA plugin, vscode-go currently doesn't have such setting.
AFAIK the setting I mentioned is the only way to force it to use modules for Go tools.

**Comment by stamblerre** [comment]
I found and fixed this issue with https://github.com/Microsoft/vscode-go/commit/c026b8c2362d7d0605c25bee24f19df88a208462, so if you follow the instructions to install the latest beta for VSCode-Go, you should be able to see the fix (https://github.com/Microsoft/vscode-go/wiki/Use-the-beta-version-of-the-latest-Go-extension). Otherwise, the fix will be out in the next release.

**Comment by nezorflame** [comment]
While it definitely works now, I believe that this new release broke `Go to definition` for the std packages like `http`, `io/ioutil` etc.
It's possible to reference functions from them and go to their definition, but once inside, `Go to definition` does nothing and doesn't work at all.

**Comment by shaodahong** [comment]
oh, if you don't import `net/http` package, you input `http.` nothing happen

**Comment by nezorflame** [comment]
@stamblerre agreed, thanks!
